### ISSUE TITLE
Possible Python2 undesired behavior

### ISSUE THREAD

**Issue by Sinclert** [diana-hep/madminer#422]
This issue aims to expose a possible undesired behavior when using the `python2_override` argument within the [madminer.core.MadMiner.run_multiple](https://github.com/diana-hep/madminer/blob/master/madminer/core/madminer.py#L784) function, and its usage in combination with the `initial_command` argument.

These arguments are used to determine what it is passed as `explicit_python_call` and `initial_command` to _downstream_ functions [generate_mg_process](https://github.com/diana-hep/madminer/blob/master/madminer/core/madminer.py#L901), [setup_mg_with_scripts](https://github.com/diana-hep/madminer/blob/master/madminer/core/madminer.py#L998) and [run_mg](https://github.com/diana-hep/madminer/blob/master/madminer/core/madminer.py#L1019) , so they _may_ be affected too.

---

### Current situation
MadMiner currently supports Python3 as a valid version to interact with the library. However, as it depends on some _under the hood_ libraries which are stuck on Python2 (like [MadGraph](https://launchpad.net/mg5amcnlo)), it provides a convenient way of running them _under the hood_ with some key arguments named `python2_override` and `initial_command`.

The `python2_override` is a flag argument which can be specified to clarify that the user wants to run the underlying Python2 stuck dependencies with the User installed Python2 distribution, instead of the System one (as the System one is better left untouched).

---

### Issue
When working with the library in a real-world environment, which dozens of projects, it is highly probable that developers are going to create virtual envs. to interact with MadMiner.

Once a virtual environment is used, the usage of `python2_override` is not enough, so that is when the `initial_command` argument becomes useful. This argument allow users to activate a custom installed env. for the _under the hood_ dependencies to use.

Example:
```python
initial_command="source venv/bin/activate",
```

However, the underlying dependencies of the library ([MadGraph](https://launchpad.net/mg5amcnlo) & friends)  **only works if that virtual environment is a Python2 environment**. This is highly inconvenient, as the users would need to define a Python3 virtual env. to interact with the library, and another virtual env. to be specified as `initial_command`.

---

### How to replicate
First create a Python3 virtual env. and install MadMiner:
```shell
$ virtualenv venv -p python3
(venv) $ pip install madminer==0.7.1
```

Then try to run a MadGraph dependent workflow, **just** specifying the `python2_override` argument as `True` when calling the [madminer.core.MadMiner.run_multiple](https://github.com/diana-hep/madminer/blob/master/madminer/core/madminer.py#L784) function. It must happens in a workflow that uses [Pythia8](http://home.thep.lu.se/Pythia/) with `reweight=ON` denoted in the MadGraph `mg_commands.dat` card.

You will get an error such as:
```shell
Command "generate_events run_01" interrupted with error:
ImportError : No module named allmatrix2py
```

Turns out that the issue has nothing to do with a missing `allmatrix2py` file, but with the installation of **Python2 Numpy** and its provided `f2py` binary, as [the answer to this MadGraph question suggest](https://answers.launchpad.net/mg5amcnlo/+question/664366) **(VIP)**.

---
### Current solution
The only way to use a Python3 virtual env. as a valid virtual env. to fully operate with MadMiner, is to specify the following values when interacting with the _under the hood_ dependencies:
```python
python2_override=True,
initial_command="export PATH=<USER_INSTALLED_PYTHON2_PATH>/bin:$PATH",
```

In macOS, for instance:
```python
python2_override=True,
initial_command="export PATH=/Library/Frameworks/Python.framework/Versions/2.7/bin:$PATH",
```

**Because we need the Python2 `f2py` binary not being _replaced_ by the virtual environment Python3 `f2py` binary**.

---

### Conclusions:
1. This is not a bug _per se_, but an annoying situation to deal with, as the current interface of the library force the Python3 users to both:
- Have 2 separate virtual envs (inconvenient).
- Use both args. `python2_override` and `initial_command`, (confusing, as users could think that just providing the `python2_override=True` option would be enough).

2. I want to emphasize [this MadGraph forum reference](https://answers.launchpad.net/mg5amcnlo/+question/664366), as it was the one giving me the key to debug this nightmare.

**Comment by johannbrehmer** [comment]
Hi @Sinclert,

Thank you for this very thorough write-up of a super annoying situation that cause me quite some gray hair in the past (and I'm sure Felix as well).

However, at least on my system your conclusions are not quite true. I use a Python 3 virtual env for MadMiner and use `initial_command` to activate a Python 2 virtual env for the MadGraph steps, *without* using the `python2_override` flag. See for instance https://github.com/diana-hep/madminer/blob/master/examples/tutorial_particle_physics/2a_parton_level_analysis.ipynb. Both virtual envs have Numpy in them. That setup works smoothly for me, but I'm not 100% sure if there is any peculiarity about my conda setup or PATH that I am not thinking of right now...

I totally agree that this is far from desirable though, but I'm at a loss of what we can do short of rewriting MadGraph in Python 3. Do you have an idea?

Thanks again,
Johann

**Comment by Sinclert** [comment]
Thanks for the quick reply.

Yes, you are right: having 2 virtual environments (one with Python2, another with Python3), and just specifying `initial_command` as a way to activate the Python2 env works.

However, I feel that the `python2_override` is hardly usable as it is right now, as it always need to be accompanied with a `initial_command` PATH setting. In other words, it is not usable just by itself.

My proposal is to add a condition on the [madminer.core.MadMiner.run_multiple](https://github.com/diana-hep/madminer/blob/master/madminer/core/madminer.py#L784) function, that adds the User-install Python2 binaries to the PATH. Example:

```python
if python2_override and initial_command is None:
    binary_path = os.popen("command -v python2.7").read().strip()
    binary_folder = os.path.dirname(os.path.realpath(binary_path))
    initial_command = "export PATH={}:$PATH".format(binary_folder)
```

This way, only when the `python2_override` flag arguments is used, and the `initial_command` value is empty, the binaries of the User-installed Python2 distribution are added to the **beginning** of the PATH.

**Comment by johannbrehmer** [comment]
That's a good idea, I'm totally on board. Also, maybe we should copy your issue description into the relevant docstring.

Do you maybe want to open a PR?

**Comment by Sinclert** [comment]
Ok, I will open a PR referencing this issue 👍 

Before that, I will spend some time exploring the _downstream_ functions ([generate_mg_process](https://github.com/diana-hep/madminer/blob/master/madminer/core/madminer.py#L901), [setup_mg_with_scripts](https://github.com/diana-hep/madminer/blob/master/madminer/core/madminer.py#L998) and [run_mg](https://github.com/diana-hep/madminer/blob/master/madminer/core/madminer.py#L1019)) just in case there is some edge case we must consider.

**Comment by johannbrehmer** [comment]
Solved in fd737c2b3b50ace06d8fd2cb337792892ad7d103
